server/socketManager.js

```diff
@@ -252,6 +221,211 @@ class SocketManager extends Server {
+      ws.destroy();
+      return;
+    }
+
+    let raw, addr;
+    try {
+      raw = IP.toBuffer(host);
+      addr = IP.toString(raw);
+    } catch (e) {
+      this.proxyLog(
+        'error',
+        'Client gave a bad host: %s (%s).',
+        host,
+        state.host
+      );
+      ws.fire('tcp error', {
+        message: 'EHOSTUNREACH',
+        code: 'EHOSTUNREACH'
+      });
+      ws.destroy();
+      return;
+    }
+
+    if (!IP.isRoutable(raw) || IP.isOnion(raw)) {
+      this.proxyLog(
+        'warning',
+        'Client is trying to connect to a bad ip: %s (%s).',
+        addr,
+        state.host
+      );
+      ws.fire('tcp error', {
+        message: 'ENETUNREACH',
+        code: 'ENETUNREACH'
+      });
+      ws.destroy();
+      return;
+    }
+
+    if (!this.ports.has(port)) {
+      this.proxyLog(
+        'warning',
+        'Client is connecting to non-whitelist port (%s).',
+        state.host
+      );
+      ws.fire('tcp error', {
+        message: 'ENETUNREACH',
+        code: 'ENETUNREACH'
+      });
+      ws.destroy();
+      return;
+    }
+
+    let socket;
+    try {
+      socket = state.connect(
+        port,
+        addr
+      );
+      this.proxyLog(
+        'info',
+        'Connecting to %s (%s).',
+        state.remoteHost,
+        state.host
+      );
+    } catch (e) {
+      this.proxyLog('error', e.message);
+      this.proxyLog('info', 'Closing %s (%s).', state.remoteHost, state.host);
+      ws.fire('tcp error', {
+        message: 'ENETUNREACH',
+        code: 'ENETUNREACH'
+      });
+      ws.destroy();
+      return;
+    }
+
+    socket.on('connect', () => {
+      ws.fire('tcp connect', socket.remoteAddress, socket.remotePort);
+    });
+
+    socket.on('data', data => {
+      ws.fire('tcp data', data.toString('hex'));
+    });
+
+    socket.on('error', err => {
+      ws.fire('tcp error', {
+        message: err.message,
+        code: err.code || null
+      });
+    });
+
+    socket.on('timeout', () => {
+      ws.fire('tcp timeout');
+    });
+
+    socket.on('close', () => {
+      this.proxyLog('info', 'Closing %s (%s).', state.remoteHost, state.host);
+      ws.fire('tcp close');
+      ws.destroy();
+    });
+
+    ws.bind('tcp data', data => {
+      if (typeof data !== 'string') return;
+      socket.write(Buffer.from(data, 'hex'));
+    });
+
+    ws.bind('tcp keep alive', (enable, delay) => {
+      socket.setKeepAlive(enable, delay);
+    });
+
+    ws.bind('tcp no delay', enable => {
+      socket.setNoDelay(enable);
+    });
+
+    ws.bind('tcp set timeout', timeout => {
+      socket.setTimeout(timeout);
+    });
+
+    ws.bind('tcp pause', () => {
+      socket.pause();
+    });
+
+    ws.bind('tcp resume', () => {
+      socket.resume();
+    });
+
+    ws.on('disconnect', () => {
+      socket.destroy();
+    });
+  }
+
+  proxyLog(level, _message, ...args) {
+    const message = `(wsproxy) ${_message}`;
+    this.logger[level](message, ...args);
```

Comment:
Since `context` is already set for the SocketManager I added this custom logger that just prepends a subsidiary context to indicate that these messages are specifically for tcp proxied messages.

Reply:
👍